pkg/cmd/infra/router/template.go

```diff
@@ -2,87 +2,163 @@ package router
-		return nil, errors.New("Reload script must be specified")
+	cmd.AddCommand(version.NewVersionCommand(name))
+
+	flag := cmd.Flags()
+	options.Config.Bind(flag)
+	options.TemplateRouter.Bind(flag)
+	options.RouterSelection.Bind(flag)
+
+	return cmd
+}
+
+func (o *TemplateRouterOptions) Complete() error {
+	routerSvcName := util.Env("ROUTER_SERVICE_NAME", "")
+	routerSvcNamespace := util.Env("ROUTER_SERVICE_NAMESPACE", "")
+	if len(routerSvcName) > 0 {
+		if len(routerSvcNamespace) == 0 {
+			return fmt.Errorf("ROUTER_SERVICE_NAMESPACE is required when ROUTER_SERVICE_NAME is specified")
+		}
+		o.RouterService = &ktypes.NamespacedName{
+			Namespace: routerSvcNamespace,
+			Name:      routerSvcName,
+		}
 	}
 
-	statsPort := 0
-	var err error = nil
-	if cfg.TemplateRouterConfig.StatsPort != "" {
-		statsPort, err = strconv.Atoi(cfg.TemplateRouterConfig.StatsPort)
+	if len(o.StatsPortString) > 0 {
+		statsPort, err := strconv.Atoi(o.StatsPortString)
 		if err != nil {
-			return nil, errors.New("Invalid stats port")
+			return fmt.Errorf("stat port is not valid: %v", err)
 		}
+		o.StatsPort = statsPort
 	}
 
-	routerSvcNamespace := util.Env("ROUTER_SERVICE_NAMESPACE", "")
-	routerSvcName := util.Env("ROUTER_SERVICE_NAME", "")
-	cfg.TemplateRouterConfig.RouterService = ktypes.NamespacedName{
-		Namespace: routerSvcNamespace,
-		Name:      routerSvcName,
+	return o.RouterSelection.Complete()
+}
+
+func (o *TemplateRouterOptions) Validate() error {
+	if len(o.TemplateFile) == 0 {
+		return errors.New("template file must be specified")
+	}
+
+	if len(o.ReloadScript) == 0 {
+		return errors.New("reload script must be specified")
+	}
+	return nil
+}
+
+// Run launches a template router using the provided options. It never exits.
+func (o *TemplateRouterOptions) Run() error {
+	pluginCfg := templateplugin.TemplatePluginConfig{
+		WorkingDir:         o.WorkingDir,
+		TemplatePath:       o.TemplateFile,
+		ReloadScriptPath:   o.ReloadScript,
+		DefaultCertificate: o.DefaultCertificate,
+		StatsPort:          o.StatsPort,
+		StatsUsername:      o.StatsUsername,
+		StatsPassword:      o.StatsPassword,
+		PeerService:        o.RouterService,
+	}
+
+	plugin, err := templateplugin.NewTemplatePlugin(pluginCfg)
+	if err != nil {
+		return err
 	}
 
-	templatePluginCfg := templateplugin.TemplatePluginConfig{
-		TemplatePath:       cfg.TemplateRouterConfig.TemplateFile,
-		ReloadScriptPath:   cfg.TemplateRouterConfig.ReloadScript,
-		DefaultCertificate: cfg.DefaultCertificate,
-		StatsPort:          statsPort,
-		StatsUsername:      cfg.TemplateRouterConfig.StatsUsername,
-		StatsPassword:      cfg.TemplateRouterConfig.StatsPassword,
-		PeerService:        cfg.TemplateRouterConfig.RouterService,
+	oc, kc, err := o.Config.Clients()
+	if err != nil {
+		return err
 	}
-	return templateplugin.NewTemplatePlugin(templatePluginCfg)
+
+	factory := o.RouterSelection.NewFactory(oc, kc)
+	controller := factory.Create(plugin)
+	controller.Run()
+
+	proc.StartReaper()
```

Comment:
I don't think the controller's scope of responsibility is managing child
processes.  Since this is a very process specific thing it's more
appropriate for a main method.

On Wed, Sep 2, 2015 at 2:30 PM, Ram Ranganathan <notifications@github.com>
wrote:

> In pkg/cmd/infra/router/template.go
> <https://github.com/openshift/origin/pull/4421#discussion_r38567160>:
>
> >  	}
> > -	return templateplugin.NewTemplatePlugin(templatePluginCfg)
> > +
> > +	factory := o.RouterSelection.NewFactory(oc, kc)
> > +	controller := factory.Create(plugin)
> > +	controller.Run()
> > +
> > +	proc.StartReaper()
>
> Since this is used in the f5 code as well (well generically to reap
> processes if pid 1) - might be better to move the proc.StartReaper() into
> the controller.Run() method.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4421/files#r38567160>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift